### PR TITLE
Add user-friendly warning messages for `returns` contract

### PR DESCRIPTION
**Pull request by gabryon99**
This PR uses the `SourceRole` information embedded in Viper's AST nodes to generate user-friendly messages. The `DiagnosticReporter` in the context of `VerifierErrorInterpreter` object can report `VerifierError` giving them an interpretation according to the input error style.

### CODE REVIEW

File: plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/ViperPoweredDeclarationChecker.kt [hunk 1]
```diff
@@ -61,12 +52,11 @@ class ViperPoweredDeclarationChecker(private val session: FirSession, private va
             }
 
             val verifier = Verifier()
-
             val onFailure = { err: VerifierError ->
-                val source = err.position.unwrapOr {
-                    declaration.source
+                val source = err.position.unwrapOr { declaration.source }
+                with(VerifierErrorInterpreter) {
+                    reporter.reportOn(source, err, config.errorStyle, context)
```
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/ViperPoweredDeclarationChecker.kt:58]: I could change the name to `reportVerifierError`. WDYT?

File: plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        ErrorStyle.USER_FRIENDLY -> reportVerificationErrorUserFriendly(source, error, context)
+        ErrorStyle.ORIGINAL_VIPER -> reportVerificationErrorOriginalViper(source, error, context)
+        ErrorStyle.BOTH -> {
+            reportVerificationErrorUserFriendly(source, error, context)
+            reportVerificationErrorOriginalViper(source, error, context)
+        }
+    }
+
+    private fun DiagnosticReporter.reportConsistencyError(source: KtSourceElement?, error: ConsistencyError, context: CheckerContext) {
+        reportOn(source, PluginErrors.INTERNAL_ERROR, error.msg, context)
```
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt:60]: Could you please clarify this comment?

[the remark this answers is not in the document]
Comment (jesyspa) [reply plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt:65]: `INTERNAL_ERROR` currently has `DECLARATION_NAME` as source error positioning strategy.  This is nice when the source is the function declaration, but it means that you shouldn't use it with arbitrary source positions.  I'm not sure whether there's a way to pass a specific strategy to `reportOn` for specific cases, but something along those lines would be good.

[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt]: In [this](https://github.com/jesyspa/kotlin/pull/172/commits/f1954b8c38b11d8ac7b5aafcb058de90b68a6abc) I add a check that verify if the source's origin is a function declaration or not. The function `reprtOn` can accept an optional parameter for the position strategy.

File: plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.jetbrains.kotlin.diagnostics.SourceElementPositioningStrategies
+import org.jetbrains.kotlin.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.formver.ErrorStyle
+import org.jetbrains.kotlin.formver.PluginErrors
+import org.jetbrains.kotlin.formver.embeddings.SourceRole
+import org.jetbrains.kotlin.formver.viper.errors.*
+
+class VerifierErrorInterpreter {
+    private var alreadyReported = false
```
Comment (jesyspa) [inline plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt:20]: This variable doesn't quite capture what it says it does: you only set it for original viper errors, not for all errors.

File: plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        reportOn(source, PluginErrors.INTERNAL_ERROR, error.msg, context, positioningStrategy = positionStrategy)
+    }
+
+    fun DiagnosticReporter.reportVerifierError(
+        source: KtSourceElement?,
+        error: VerifierError,
+        errorStyle: ErrorStyle,
+        context: CheckerContext,
+    ) =
```
Comment (jesyspa) [inline plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt:80]: Is the formatting here correct?

File: plugins/formal-verification/formver.viper/src/org/jetbrains/kotlin/formver/viper/errors/VerificationError.kt [hunk 5]
```diff
@@ -45,4 +47,13 @@ object ErrorAdapter {
...
         }
     }
-}
\ No newline at end of file
+}
+
+/**
+ * Given a verification error, find embedded extra information of type `I` in the
+ * error's offending nodes. If such information is not found, return `default` as value.
+ */
```
Comment (jesyspa) [inline plugins/formal-verification/formver.viper/src/org/jetbrains/kotlin/formver/viper/errors/VerificationError.kt:55]: This seems to be neither a full tree search nor just an immediate lookup.  Could you document what it is looking for?  Are calls like `_2()` always safe to do, or could this result in some kind of out-of-bounds exception?

File: plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    ) = when (errorStyle) {
+        ErrorStyle.USER_FRIENDLY -> reportVerificationErrorUserFriendly(source, error, context)
+        ErrorStyle.ORIGINAL_VIPER -> reportVerificationErrorOriginalViper(source, error, context)
+        ErrorStyle.BOTH -> {
+            reportVerificationErrorUserFriendly(source, error, context)
+            reportVerificationErrorOriginalViper(source, error, context)
+        }
+    }
+
+    private fun DiagnosticReporter.reportConsistencyError(source: KtSourceElement?, error: ConsistencyError, context: CheckerContext) {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt:66]: Is the `CONSISTENCY_ERROR` diagnostic still used?

**Review by jesyspa — APPROVED** [review]

File: plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt (hunk 6 above)
Comment (gabryon99) [reply plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt:66]: No, we do not use it anymore. I will remove it!

File: plugins/formal-verification/formver.viper/src/org/jetbrains/kotlin/formver/viper/errors/VerificationError.kt (hunk 5 above)
Comment (gabryon99) [reply plugins/formal-verification/formver.viper/src/org/jetbrains/kotlin/formver/viper/errors/VerificationError.kt:55]: > This seems to be neither a full tree search nor just an immediate lookup. Could you document what it is looking for?

Sure! I will write a short summary.

> Are calls like _2() always safe to do, or could this result in some kind of out-of-bounds exception?

Yes, they are calls to access Scala Tuples`N`'s elements (in this case `Tuple3` exposes three accessor methods `_1, _2, _3`).

**Pull request merged**